src/main/java/org/sunbird/hubservices/daoimpl/GraphDao.java

```diff
@@ -0,0 +1,215 @@
+
+            Map<String, Object> props = new HashMap<>();
+            props.put("id", UUID);
+            //props.put( "name", "tag-"+UUID );
+
+            Map<String, Object> params = new HashMap<>();
+            params.put("props", props);
+            params.put("uuid", UUID);
+
+            StringBuilder queryBuilder = new StringBuilder();
+            queryBuilder.append("MATCH (n:").append(nodeType).append(") WHERE n.id = $uuid RETURN n");
+            Statement statement = new Statement(queryBuilder.toString(), params);
+
+            StatementResult result = transaction.run(statement);
+            int recordSize = result.list().size();
+            logger.info("records {} ", recordSize);
+            result.consume();
+
+            if (recordSize == 0) {
+                queryBuilder = new StringBuilder();
+                queryBuilder.append("CREATE (n:").append(nodeType).append(") SET n = $props RETURN n");
+                statement = new Statement(queryBuilder.toString(), params);
+                result = transaction.run(statement);
+                result.consume();
+                transaction.commitAsync().toCompletableFuture();
+                logger.info("user node with id {} created successfully ", UUID);
+            } else {
+                logger.info("user node with id {} exist ", UUID);
+
+            }
+            transaction.close();
+
+        } catch (ClientException e) {
+            transaction.rollbackAsync().toCompletableFuture();
+            logger.error("user node upsertion failed : {}", e);
+        }
+        session.close();
+    }
+
+    @Override
+    public void upsertRelation(String fromUUID, String toUUID, String propertyName, String propertyValue) throws ClientException {
+        Session session = neo4jDriver.session();
+        Transaction transaction = session.beginTransaction();
+
+        try {
+            if (StringUtils.isBlank(fromUUID) || StringUtils.isBlank(toUUID) || StringUtils.isBlank(propertyName) || StringUtils.isBlank(propertyValue)) {
+                throw new ClientException("UUIDs or property cannot be empty");
+            }
+
+            Map<String, Object> parameters = new HashMap<>();
+            parameters.put("fromUUID", fromUUID);
+            parameters.put("toUUID", toUUID);
+            parameters.put("propertyValue", propertyValue);
+
+            StringBuilder query = new StringBuilder();
+            query.append("MATCH (n:").append(nodeType).append("), (n1:").append(nodeType)
+                    .append(") WHERE n.id = $fromUUID AND n1.id = $toUUID RETURN n,n1");
+            Statement statement = new Statement(query.toString(), parameters);
+            StatementResult result = transaction.run(statement);
+            int recordSize = result.list().size();
+            result.consume();
+            if (recordSize == 0) { // nodes has no relation
+                throw new NoSuchRecordException("users with toUUID {} or fromUUID {} not found");
```

Comment:
Should we look at adding the nodes if not found during an upsert implementation.